aws/logs_monitoring/lambda_function.py

```diff
@@ -61,78 +72,198 @@
-    def __init__(self, host, port, ddApiKey):
-        self.host = host
-        self.port = port
-        self.api_key = ddApiKey
-        self._sock = None
+class RetriableException(Exception):
+    pass
 
-    def _connect(self):
-        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-        s = ssl.wrap_socket(s)
-        s.connect((self.host, self.port))
-        return s
 
-    def safe_submit_log(self, log, metadata):
-        try:
-            self.send_entry(log, metadata)
-        except Exception as e:
-            # retry once
-            if self._sock:
-                # make sure we don't keep old connections open
-                self._sock.close()
-            self._sock = self._connect()
-            self.send_entry(log)
-        return self
+class ScrubbingException(Exception):
+    pass
 
-    def send_entry(self, log_entry, metadata):
-        # The log_entry can only be a string or a dict
-        if isinstance(log_entry, str):
-            log_entry = {"message": log_entry}
-        elif not isinstance(log_entry, dict):
-            raise Exception(
-                "Cannot send the entry as it must be either a string or a dict. Provided entry: "
-                + str(log_entry)
-            )
 
-        # Merge with metadata
-        log_entry = merge_dicts(log_entry, metadata)
+class DatadogClient(object):
+    """
+    Client that implements a linear retrying logic to send a batch of logs.
+    """
 
-        # Send to Datadog
-        str_entry = json.dumps(log_entry)
+    def __init__(self, client, max_retries=3, backoff=1):
+        self._client = client
+        self._max_retries = max_retries
+        self._backoff = backoff
 
-        # Scrub ip addresses if activated
-        if is_ipscrubbing:
+    def send(self, logs, metadata):
+        retry = 0
+        while retry <= self._max_retries:
+            if retry > 0:
+                time.sleep(self._backoff)
             try:
-                str_entry = ip_regex.sub("xxx.xxx.xxx.xx", str_entry)
-            except Exception as e:
-                print(
-                    "Unexpected exception while scrubbing logs: {} for event {}".format(
-                        str(e), str_entry
-                    )
-                )
-
-        # For debugging purpose uncomment the following line
-        # print(str_entry)
-        prefix = "%s " % self.api_key
-        return self._sock.send((prefix + str_entry + "\n").encode("UTF-8"))
+                self._client.send(logs, metadata)
+            except RetriableException as e:
+                retry += 1
+                continue
+            return
+        raise Exception("max number of retries reached: {}".format(self._max_retries))
 
     def __enter__(self):
-        self._sock = self._connect()
+        self._client.connect()
         return self
 
     def __exit__(self, ex_type, ex_value, traceback):
+        self._client.close()
+        return self
+
+
+class DatadogTCPClient(object):
```

Comment:
I would avoid refactoring the TCP client, I'd keep the logic untouched if possible. 
Why? Because by refactoring it we risk introducing bugs, while we simply want to deprecate it. Also when we'll touch common logic we'll need to test twice, which we might not do so we might introduce regressions

Reply:
Agreed but changes are safe, I am quite confident that nothing will break here and the goal is to fully switch to HTTP, TCP will only be used as a backup if needed.